scripts/functions-deploy-tests/tests.ts

```diff
@@ -0,0 +1,330 @@
+
+async function setOpts(opts: Opts) {
+  let stmt = "";
+  for (const [name, opt] of Object.entries(opts)) {
+    if (opt) {
+      stmt += `export const ${name} = ${JSON.stringify(opt)};\n`;
+    }
+  }
+  await fs.writeFile(path.join(FUNCTIONS_DIR, "options.js"), stmt);
+}
+
+async function listFns(stripId = "dvtuqrxfjr"): Promise<Record<string, Endpoint>> {
+  const result = await cli.exec("functions:list", FIREBASE_PROJECT, ["--json"], __dirname, false);
+  const output = JSON.parse(result.stdout);
+
+  const eps: Record<string, Endpoint> = {};
+  for (const ep of output.result as Endpoint[]) {
+    const id = ep.id.replace(`${stripId}-`, "");
+    if (ep.id === id) {
+      // This endpoint is from another run. Ignore.
+      continue;
+    }
+    eps[id] = ep;
+  }
+  return eps;
+}
+
+describe("firebase deploy", function (this) {
+  this.timeout(1000_000);
+
+  const RUN_ID = genRandomId();
+  console.log(`TEST RUN: ${RUN_ID}`);
+
+  before(async () => {
+    expect(FIREBASE_PROJECT).to.not.be.empty;
+
+    // write up index.js to import trigger definition using unique group identifier.
+    // All exported functions will have name {hash}-{trigger} e.g. 'abcdefg-v1storage'.
+    await fs.writeFile(
+      path.join(FUNCTIONS_DIR, "index.js"),
+      `export * as ${RUN_ID} from "./fns.js";`
+    );
+  });
+
+  after(async () => {
+    await fs.unlink(path.join(FUNCTIONS_DIR, "index.js"));
+  });
+
+  it("deploys functions with runtime options", async () => {
+    const opts: Opts = {
+      v1Opts: {
+        memory: "128MB",
+        maxInstances: 42,
+        timeoutSeconds: 42,
+      },
+      v2Opts: {
+        memory: "128MiB",
+        maxInstances: 42,
+        timeoutSeconds: 42,
+        // TODO: Re-enable once https://github.com/firebase/firebase-tools/issues/4679 is fixed.
+        // cpu: 2,
+        concurrency: 42,
+      },
+      v1TqOpts: {
+        retryConfig: {
+          maxAttempts: 42,
+          maxRetrySeconds: 42,
+          maxBackoffSeconds: 42,
+          maxDoublings: 42,
+          minBackoffSeconds: 42,
+        },
+        rateLimits: {
+          maxDispatchesPerSecond: 42,
+          maxConcurrentDispatches: 42,
+        },
+      },
+      v2TqOpts: {
+        retryConfig: {
+          maxAttempts: 42,
+          maxRetrySeconds: 42,
+          maxBackoffSeconds: 42,
+          maxDoublings: 42,
+          minBackoffSeconds: 42,
+        },
+        rateLimits: {
+          maxDispatchesPerSecond: 42,
+          maxConcurrentDispatches: 42,
+        },
+      },
+      v1IdpOpts: {
+        blockingOptions: {
+          idToken: true,
+          refreshToken: true,
+          accessToken: false,
+        },
+      },
+      v2IdpOpts: {
+        idToken: true,
+        refreshToken: true,
+        accessToken: true,
+      },
+    };
+    await setOpts(opts);
+
+    const result = await cli.exec(
+      "deploy",
+      FIREBASE_PROJECT,
+      ["--only", "functions", "--non-interactive", "--force"],
```

Comment:
I think it'd be helpful to have a command-line flag that causes this to be run with the --debug flag also.

Reply:
Good suggestion. I found myself doing that also 🤔.

Reply:
Put together a way to set env var. Hacky, but works:

```
$  GCLOUD_PROJECT=${PROJECT_ID} FIREBASE_DEBUG=true npm run test:functions-deploy
```